### ISSUE TITLE
tendermint-rs should expose traits for crypto operations

### ISSUE THREAD

**Issue by seunlanlege** [cosmos/ibc-rs#18]
## Problem Definition

In order for this library to be used in smart contract and blockchain environments, as i believe is the intention, crypto (sign, hash, verify) operations should be instead delegated to trait implementations. Reason is crypto operations are very expensive and slow in wasm. And for this reason the substrate runtime exposes host functions for hashing:

https://github.com/paritytech/substrate/blob/9461b2de04210c6c193726a745c3ec6552b4ce9f/primitives/io/src/lib.rs#L993-L1033

as well as signature verification and digital signature schemes: 

https://github.com/paritytech/substrate/blob/9461b2de04210c6c193726a745c3ec6552b4ce9f/primitives/io/src/lib.rs#L616-L989

Using a trait will allow delegating these operations to pre-compiles (aka host functions) 

cc @adizere

**Comment by adizere** [comment]
This seems important, thank you for flagging Seun.

Let's clarify the context:
Do I understand correctly this feature is for the `ibc` [crate](https://github.com/informalsystems/ibc-rs/tree/master/modules) implementation? In this crate we do minimal cryptographic operations, limited to proof verification, all of which [are delegated to the ICS23 crate](https://github.com/informalsystems/ibc-rs/blob/36240d0fb9a344e23e76ec8996ee09f66f86c82f/modules/src/core/ics23_commitment/merkle.rs#L8-L10). So not sure I understand.

(Or is it for the relayer library? The relayer is not supposed to run on-chain.)

Can you clarify exactly which method from `ibc-rs` or `tendermint-rs` should be delegated to trait implementations?

**Comment by seunlanlege** [comment]
> Do I understand correctly this feature is for the ibc [crate](https://github.com/informalsystems/ibc-rs/tree/master/modules) implementation? In this crate we do minimal cryptographic operations, limited to proof verification, all of which [are delegated to the ICS23 crate](https://github.com/informalsystems/ibc-rs/blob/36240d0fb9a344e23e76ec8996ee09f66f86c82f/modules/src/core/ics23_commitment/merkle.rs#L8-L10).

Yes this is specifically for the `ibc` crate. So what this means is that the `ics23` crate will have to be generic over the crypto operations it uses and expose that to the `ibc` crate, so that in-turn can be left up to downstream implementations to implement however they like.

But this also goes for the `tendermint-rs` crate which also does crypto operations like signature verification and hashing.

**Comment by adizere** [comment]
> Yes this is specifically for the `ibc` crate. So what this means is that the `ics23` crate will have to be generic over the crypto operations it uses and expose that to the `ibc` crate, so that in-turn can be left up to downstream implementations to implement however they like.

I see. Then I think we'll need to define these requirements in an issue over at https://github.com/confio/ics23. I had a look over at the library and the most compute-intensive method seem to be [`do_hash`](https://github.com/confio/ics23/blob/8ea0b239740068dd004fe4a42a1d861cd607611c/rust/src/ops.rs#L35), so this might be the method we want to make generic. We should check with the maintainer of ics23 @ethanfrey, however, since my knowledge of that codebase is very limited.

Once the required changes land in `ics23` we'll add support for that here. Do you agree with that approach and is it clear, or do you think there are specific changes we could do directly in `ibc` crate?

**Comment by adizere** [comment]
> But this also goes for the `tendermint-rs` crate which also does crypto operations like signature verification and hashing.

Good point. I think we use `tendermint-rs` in the implementation of ICS07 verification methods. There may be others, but I was able to identify this one specifically:

https://github.com/informalsystems/ibc-rs/blob/b44ade9f88c7faf4d0b6f9e672dbef08d2ab3817/modules/src/clients/ics07_tendermint/client_def.rs#L113

@seunlanlege We will need from you the exact list of crypto methods (from `verifier.verify`) that you would like to be delegated. Then we can work with the `tendermint-rs` team to adapt the library to your needs and see how we can expose those methods in a trait.

**Comment by seunlanlege** [comment]
> Once the required changes land in ics23 we'll add support for that here. Do you agree with that approach and is it clear, or do you think there are specific changes we could do directly in ibc crate?

Yes this is fine, although some hash functions unfortunately aren't supported by substrate yet like `Ripemd160`

> @seunlanlege We will need from you the exact list of crypto methods (from verifier.verify) that you would like to be delegated. Then we can work with the tendermint-rs team to adapt the library to your needs and see how we can expose those methods in a trait.

honestly not super familiar with the tendermint-rs lib, so this was all i could find:

https://github.com/informalsystems/tendermint-rs/blob/7a8d681027b92a27174a3fd5c6e893dc35f42c31/tendermint/src/public_key.rs#L197-L226

But i suspect there maybe more that i'm missing

**Comment by plafer** [comment]
@seunlanlege @blasrodri I believe this is what we discussed recently, that you already implemented this in your fork, and that you will eventually open a new issue/PR here that addresses it?

I will close this as stale for now; please open a new issue with specific details of what needs to be done (and possibly a PR if you already did it on your fork).
